File: chapter_3_stacks_and_queues/js/queue.js
/* first in, first out */
class Queue {
  constructor() {
    this.queue = []
    this.length = 0
  }

  enqueue(value) {
    this.queue.push(value)
    this.length++
  }

  dequeue() {
    this.length--
    return this.queue.shift()
  }

  isEmpty() {
    return this.length == 0
  }

  peek() {
    return this.queue[0]
  }
}

module.exports = {
  Queue
}

const q = new Queue()
const arr = [4, 5, 2, 8, 1, 9, 6, 7, 3, 0]
for (let i of arr) q.enqueue(i)
console.log(q)
/*
Queue { queue: [ 4, 5, 2, 8, 1, 9, 6, 7, 3, 0 ], length: 10 }
*/
console.log(q.dequeue()) // 4
console.log(q)
/*
Queue { queue: [ 5, 2, 8, 1, 9, 6, 7, 3, 0 ], length: 9 } 
*/
console.log(q.peek()) // 5

File: chapter_3_stacks_and_queues/python/queue.py
# first in, first out
class Queue():
    def __init__(self):
        self.queue = []
        self.length = 0

    def enqueue(self, value):
        self.queue.append(value)
        self.length += 1

    def dequeue(self):
        self.length -= 1
        return self.queue.pop(0)

    def peek(self):
        return self.queue[0]

    def is_empty(self):
        return self.length == 0


if __name__ == 'main':
    q = Queue()
    for i in range(5):
        q.enqueue(i)
    print(q.queue)  # [0, 1, 2, 3, 4]
    print(q.dequeue())  # 0
    print(q.peek())  # 1
